File: go/benchmark_test.go
package aps_test

import (
	"crypto/ed25519"
	"fmt"
	"testing"

	"github.com/cezexPL/agent-passport-standard/go/crypto"
	"github.com/cezexPL/agent-passport-standard/go/passport"
)

var sinkString string
var sinkBytes []byte
var sinkBool bool

func BenchmarkCanonicalizeJSON(b *testing.B) {
	obj := map[string]interface{}{
		"z_last": "value",
		"a_first": 42,
		"m_middle": []interface{}{"x", "y"},
		"nested": map[string]interface{}{"b": 2, "a": 1},
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		data, _ := crypto.CanonicalizeJSON(obj)
		sinkBytes = data
	}
}

func BenchmarkKeccak256(b *testing.B) {
	data := []byte("The quick brown fox jumps over the lazy dog")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		sinkString = crypto.Keccak256(data)
	}
}

func BenchmarkKeccak256Large(b *testing.B) {
	data := make([]byte, 1024*1024) // 1MB
	for i := range data {
		data[i] = byte(i % 256)
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		sinkString = crypto.Keccak256(data)
	}
}

func BenchmarkEd25519Sign(b *testing.B) {
	_, priv, _ := crypto.GenerateKeyPair()
	data := []byte("benchmark signing payload")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		sinkString = crypto.Ed25519Sign(priv, data)
	}
}

func BenchmarkEd25519Verify(b *testing.B) {
	pub, priv, _ := crypto.GenerateKeyPair()
	data := []byte("benchmark verify payload")
	sig := crypto.Ed25519Sign(priv, data)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		sinkBool, _ = crypto.Ed25519Verify(pub, data, sig)
	}
}

func BenchmarkPassportCreate(b *testing.B) {
	cfg := passport.Config{
		ID:        "did:key:z6MkBenchTest1234567890abcdefghijklmnop",
		PublicKey: "abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890",
		OwnerDID:  "did:key:z6MkBenchOwner1234567890abcdefghijklmno",
		Skills: []passport.Skill{{
			Name: "go-backend", Version: "1.0.0",
			Description: "Go backend development", Capabilities: []string{"code_write"},
			Hash: "0x0000000000000000000000000000000000000000000000000000000000000000",
		}},
		Soul: passport.Soul{
			Personality: "Efficient", WorkStyle: "Systematic",
			Constraints: []string{"no-external-calls"},
			Hash: "0x0000000000000000000000000000000000000000000000000000000000000000",
		},
		Policies: passport.Policies{
			PolicySetHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
			Summary:       []string{"read-only"},
		},
		Lineage: passport.Lineage{Kind: "original", Parents: []string{}, Generation: 0},
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = passport.New(cfg)
	}
}

func BenchmarkPassportSignVerify(b *testing.B) {
	pub, priv, _ := crypto.GenerateKeyPair()
	cfg := passport.Config{
		ID:        "did:key:z6MkBenchTest1234567890abcdefghijklmnop",
		PublicKey: fmt.Sprintf("%x", pub),
		OwnerDID:  "did:key:z6MkBenchOwner1234567890abcdefghijklmno",
		Skills: []passport.Skill{{
			Name: "go-backend", Version: "1.0.0",
			Description: "Go backend development", Capabilities: []string{"code_write"},
			Hash: "0x0000000000000000000000000000000000000000000000000000000000000000",
		}},
		Soul: passport.Soul{
			Personality: "Efficient", WorkStyle: "Systematic",
			Constraints: []string{"no-external-calls"},
			Hash: "0x0000000000000000000000000000000000000000000000000000000000000000",
		},
		Policies: passport.Policies{
			PolicySetHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
			Summary:       []string{"read-only"},
		},
		Lineage: passport.Lineage{Kind: "original", Parents: []string{}, Generation: 0},
	}
	p, _ := passport.New(cfg)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = p.Sign(priv)
		_, _ = p.Verify(ed25519.PublicKey(pub))
	}
}

func BenchmarkMerkleTree1000(b *testing.B) {
	leaves := make([]string, 1000)
	for i := range leaves {
		leaves[i] = crypto.Keccak256([]byte(fmt.Sprintf("leaf-%d", i)))
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		mt := crypto.NewMerkleTree(leaves)
		sinkString = mt.Root()
	}
}

func BenchmarkSnapshotHash(b *testing.B) {
	payload := map[string]interface{}{
		"skills":   map[string]interface{}{"entries": []interface{}{}, "frozen": false},
		"soul":     map[string]interface{}{"personality": "test", "hash": "0x00"},
		"policies": map[string]interface{}{"policy_set_hash": "0x00", "summary": []interface{}{}},
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		sinkString, _ = crypto.SnapshotHash(payload)
	}
}

File: typescript/tests/benchmark.test.ts
/**
 * Performance benchmarks for APS TypeScript SDK.
 *
 * Run with: npx vitest run tests/benchmark.test.ts
 * For vitest bench mode: npx vitest bench tests/benchmark.test.ts
 */
import { describe, it, expect } from 'vitest';
import {
  canonicalizeJson,
  keccak256,
  keccak256Bytes,
  snapshotHash,
  ed25519Sign,
  ed25519Verify,
  generateKeyPair,
  MerkleTree,
} from '../src/index.js';
import { AgentPassport } from '../src/passport.js';
import { bytesToHex } from '@noble/hashes/utils';
import type { PassportConfig } from '../src/types.js';

function makeConfig(): { cfg: PassportConfig; pub: Uint8Array; priv: Uint8Array } {
  const { publicKey, privateKey } = generateKeyPair();
  const cfg: PassportConfig = {
    id: 'did:key:z6MkBenchTest1234567890abcdefghijklmnop',
    publicKey: bytesToHex(publicKey),
    ownerDID: 'did:key:z6MkBenchOwner1234567890abcdefghijklmno',
    skills: [{
      name: 'ts-backend', version: '1.0.0',
      description: 'TypeScript backend', capabilities: ['code_write'],
      hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
    }],
    soul: {
      personality: 'Efficient', work_style: 'Systematic',
      constraints: ['no-external-calls'],
      hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
      frozen: false,
    },
    policies: {
      policy_set_hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
      summary: ['read-only'],
    },
    lineage: { kind: 'original', parents: [], generation: 0 },
  };
  return { cfg, pub: publicKey, priv: privateKey };
}

function measure(fn: () => void, iterations: number): { meanUs: number; minUs: number; maxUs: number } {
  const times: number[] = [];
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    fn();
    times.push((performance.now() - start) * 1000); // µs
  }
  const mean = times.reduce((a, b) => a + b) / times.length;
  return { meanUs: mean, minUs: Math.min(...times), maxUs: Math.max(...times) };
}

async function measureAsync(fn: () => Promise<void>, iterations: number) {
  const times: number[] = [];
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    await fn();
    times.push((performance.now() - start) * 1000);
  }
  const mean = times.reduce((a, b) => a + b) / times.length;
  return { meanUs: mean, minUs: Math.min(...times), maxUs: Math.max(...times) };
}

describe('benchmarks', () => {
  it('keccak256', () => {
    const data = new TextEncoder().encode('The quick brown fox jumps over the lazy dog');
    const r = measure(() => keccak256(data), 5000);
    console.log(`keccak256: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(1000);
  });

  it('canonicalizeJson', () => {
    const r = measure(() => {
      canonicalizeJson({ z: 'last', a: 42, m: ['x', 'y'], nested: { b: 2, a: 1 } });
    }, 5000);
    console.log(`canonicalizeJson: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(1000);
  });

  it('snapshotHash', () => {
    const r = measure(() => {
      snapshotHash({
        skills: { entries: [], frozen: false },
        soul: { personality: 'test', hash: '0x00' },
        policies: { policy_set_hash: '0x00', summary: [] },
      });
    }, 3000);
    console.log(`snapshotHash: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(5000);
  });

  it('ed25519 sign', async () => {
    const { privateKey } = generateKeyPair();
    const data = new TextEncoder().encode('bench');
    const r = await measureAsync(async () => { await ed25519Sign(privateKey, data); }, 1000);
    console.log(`ed25519Sign: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(10000);
  });

  it('ed25519 verify', async () => {
    const { publicKey, privateKey } = generateKeyPair();
    const data = new TextEncoder().encode('bench');
    const sig = await ed25519Sign(privateKey, data);
    const r = await measureAsync(async () => { await ed25519Verify(publicKey, data, sig); }, 200);
    console.log(`ed25519Verify: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(10000);
  });

  it('passport create', async () => {
    const { cfg } = makeConfig();
    const r = await measureAsync(async () => { await AgentPassport.create(cfg); }, 1000);
    console.log(`passport create: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(50000);
  });

  it('passport sign+verify', async () => {
    const { cfg, pub, priv } = makeConfig();
    const p = await AgentPassport.create(cfg);
    const r = await measureAsync(async () => {
      await p.sign(priv);
      await p.verify(pub);
    }, 500);
    console.log(`passport sign+verify: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(100000);
  });

  it('merkle tree 1000 leaves', () => {
    const leaves: string[] = [];
    for (let i = 0; i < 1000; i++) {
      leaves.push(keccak256(new TextEncoder().encode(`leaf-${i}`)));
    }
    const r = measure(() => {
      const mt = new MerkleTree(leaves);
      mt.root();
    }, 50);
    console.log(`merkle 1000: mean=${r.meanUs.toFixed(1)}µs`);
    expect(r.meanUs).toBeLessThan(5000000);
  });
});
